Refactor/location localization

Joseph-Sameh-0 opened the pull request:
<img width="304" height="135" alt="image" src="https://github.com/user-attachments/assets/ccf2c6a7-4047-4f19-b7b8-5a90eb3e8a1c" />

<br/>

<img width="469" height="908" alt="image" src="https://github.com/user-attachments/assets/8e68fb89-52f7-4ea0-be10-78e8b3568920" />

<br/>

<img width="449" height="835" alt="image" src="https://github.com/user-attachments/assets/f1df7b50-c461-4cac-9541-0da7eb193db8" />

<br/>

<img width="363" height="375" alt="image" src="https://github.com/user-attachments/assets/31a1e645-eb37-423b-a0da-0259c37f96b1" />


<img width="482" height="439" alt="image" src="https://github.com/user-attachments/assets/5cb5b5f7-fc0e-414e-bccd-f90e425c4da5" />

<br/>

<img width="462" height="225" alt="image" src="https://github.com/user-attachments/assets/00e983eb-ea43-4aa1-b857-df56e18a8fc1" />

<br/>

<img width="473" height="801" alt="image" src="https://github.com/user-attachments/assets/cdd430f4-b7b6-4815-8d4e-47e1b68abb9f" />

<br/>

<img width="465" height="826" alt="image" src="https://github.com/user-attachments/assets/5f1c546d-c3f8-4521-9db5-0942a62b0b5d" />

<br/>

<img width="318" height="332" alt="image" src="https://github.com/user-attachments/assets/6f18017e-cca8-4cdb-8a0e-85e02a591b36" />

<br/>

<img width="464" height="388" alt="image" src="https://github.com/user-attachments/assets/fbb13b6b-8c3d-4a2d-b072-120936564a9d" />

Joseph-Sameh-0 committed:
Refactor: Update location data handling and UI

This commit refactors the location data handling by:
- Introducing local JSON files for governorates and cities.
- Updating `LocationLocalDataSource` to read data from these local files.
- Modifying `LocationRepositoryImp` to use `LocationLocalDataSource` instead of remote data source for fetching governorates and cities.
- Updating `GetLocationInfoUseCase` to reflect changes in `LocationRepository`.
- Changing `LocationScreenState` and `AccountUiState` to use `Governorate` and `City` domain entities instead of strings.
- Adjusting `LocationViewModel` and `AccountViewModel` to work with the new domain entities and updated use case.
- Updating UI components (`AccountScreen`, `GovernmentBottomSheet`, `LocationContent`, `LocationCard`) to display and handle `Governorate` and `City` objects.
- Adding mappers in `LocationMapper.kt` to convert DTOs to domain entities based on language.
- Renaming `Coutry` to `Country` in `LocationRemoteDataSourceImp.kt` and `LocationService.kt`.

Joseph-Sameh-0 committed:
Merge branch 'develop' into refactor/location-localization

Joseph-Sameh-0 committed:
Refactor: Update location handling and remove unused CitiesBottomSheet

- Deleted `CitiesBottomSheet.kt` as it is no longer used.
- Updated `LocationScreenState` to use `List<City>` instead of `List<String>` for cities.
- Modified `LocationScreen` to use `governorate.name` and `city.name` for display.
- Adjusted `CustomerHomeViewModel` to fetch governorates and cities using updated use case methods and handle `Governorate` and `City` objects.
- Updated `LocationViewModel` to use `City` objects and handle city selection.
- Modified `CustomerUIState` to store `List<Governorate>` and `List<City>` for bottom sheet data.
- Updated `CustomerHomeScreen` to correctly pass and handle `Governorate` and `City` names for selection.
- Marked areas needing further attention with //TODO comments regarding ID usage.

Joseph-Sameh-0 committed:
Refactor: Remove location-related files and dependencies

- Deleted `CitiesDto.kt`, `CitiesRequest.kt`, `LocationRemoteDataSource.kt`, `LocationRemoteDataSourceImp.kt`, `LocationService.kt`, and `StatesDto.kt`.
- Removed `LocationRemoteDataSource` and `LocationRemoteDataSourceImp` from `dataModule.kt`.
- Removed `LocationService` and related Retrofit configurations from `networkModule.kt`.
- Consolidated OkHttpClient and Retrofit instances in `networkModule.kt`.

Joseph-Sameh-0 committed:
Refactor: Remove location-related files and dependencies

- Deleted `CitiesDto.kt`, `CitiesRequest.kt`, `LocationRemoteDataSource.kt`, `LocationRemoteDataSourceImp.kt`, `LocationService.kt`, and `StatesDto.kt`.
- Removed `LocationRemoteDataSource` and `LocationRemoteDataSourceImp` from `dataModule.kt`.
- Removed `LocationService` and related Retrofit configurations from `networkModule.kt`.
- Consolidated OkHttpClient and Retrofit instances in `networkModule.kt`.

Joseph-Sameh-0 committed:
Refactor: Update location handling for requests and user profiles

This commit introduces several changes to how location information is handled for service requests and user profiles:

- **Requests:**
    - `RequestService` and `RequestServiceDto` now use `governorateId` and `cityId` instead of a single `location` string.
    - `toMyJobOfferUiState` in `MyOfferCraftsmanUiState` now accepts a location string to be constructed from governorate and city names.
    - `CustomerHomeViewModel` now uses `governorateId` and `cityId` when creating requests and fetches governorate/city names for display.
    - `CustomerRequestDetailsRequestDetailsViewModel` and `CraftsmanRequestDetailsViewModel` fetch governorate and city names based on IDs to display the full location.
    - `CraftsmanHomeViewModel` now fetches and displays governorate and city names for recent and available jobs.
    - `MyJobsCraftsmanViewModel` fetches and constructs the location string for display in job lists.
- **User Profile:**
    - `CraftsmanHomeViewModel` now correctly fetches and displays the user's governorate and city name.
- **UI State:**
    - `CustomerUIState` and `BottomSheetUiState` updated to store `bottomSheetSelectedGovernmentId`, `bottomSheetSelectedGovernmentName`, `bottomSheetSelectedCityId`, and `bottomSheetSelectedCityName`.
- **Data Source:**
    - `LocationLocalDataSource` methods are now suspend functions.
- **Removed Unused Mappers:**
    - Removed `toMyJobOfferUiStateMap` and `toMyJobOfferUiStateList` from `MyOfferCraftsmanUiState`.
    - Removed `toRequestServiceUiStateMap` from `CraftsmanHomeUiState`.
    - Removed `toRequestService` mapper from `CraftsmanRequestDetailsScreenState`.

Joseph-Sameh-0 committed:
Refactor: Improve error handling and UI in messaging

- Handle null location data in `User` entity by providing an empty map as a default value.
- Adjust the placement of `sendingMessage` item in `MessageDetails` screen to ensure it's displayed correctly within the LazyColumn.

Joseph-Sameh-0 committed:
Refactor: Improve location selection in customer home

This commit refactors the location selection functionality in the customer home screen and view model.

- Updated `CustomerHomeViewModel` to use government ID instead of name when fetching cities, ensuring accurate city lists.
- Set `locationBottomSheetType` to `CITY` only if cities are successfully fetched and not empty.
- Set `locationBottomSheetType` to `GOVERNMENT` when governments are fetched.
- Ensured `government` is not null before accessing its properties to prevent potential null pointer exceptions.
- Hid the government sheet after selecting a city.
- Removed redundant logic for determining `locationBottomSheetContentType` in `CustomerHomeScreen` by relying on the value from the view model state.

Joseph-Sameh-0 committed:
Refactor: Improve location selection logic and UI state

- Disable location button when a governorate is selected and no city is chosen yet.
- Enable location button once a city is selected.
- When fetching initial user location, correctly set the selected city name and fetch cities for the selected governorate.

Joseph-Sameh-0 committed:
Merge branch 'develop' into refactor/location-localization

mohammed-akkad commented on app/src/main/java/com/paris_2/san3a/domain/entity/Location.kt:
```diff
@@ -1,7 +1,16 @@
\ No newline at end of file
+){
+    companion object {
+        fun fromJson(data: Map<String, Any>) : Location {
```
Why fromJson in domain entity?
If we use api or another data source, we need it or no?

Joseph-Sameh-0 replied:
yeah I should make it on the dto but I don't have time now to separate it to domain and dto 😞

AzizaHelmy reviewed (approved):
3aash 👏

Joseph-Sameh-0 committed:
Refactor: Enhance location formatting in job requests

Joseph-Sameh-0 committed:
Refactor: Improve location formatting and offer handling in request details

- Update location string formatting in `CraftsmanRequestDetailsViewModel` and `CustomerRequestDetailsRequestDetailsViewModel` to gracefully handle null `governorate` or `city` names.
- Ensure `offers` are sorted and mapped correctly in `CustomerRequestDetailsRequestDetailsViewModel`, providing an empty map if `offers` is null.
- Streamline `acceptOfferUseCase` parameters in `CustomerRequestDetailsRequestDetailsViewModel`.

Joseph-Sameh-0 committed:
Refactor: Improve address formatting in MyOfferCraftsmanUiState
